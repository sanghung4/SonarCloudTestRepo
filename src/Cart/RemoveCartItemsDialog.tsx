import {
  Box,
  Button,
  Dialog,
  DialogContent,
  DialogTitle,
  IconButton,
  Grid,
  Typography,
  useScreenSize
} from '@dialexa/reece-component-library';
import { useTranslation } from 'react-i18next';

import { CloseIcon } from 'icons';
import { useCartContext } from 'providers/CartProvider';

type Props = {
  open: boolean;
  onClose: (save: boolean) => void;
};

function RemoveCartItemsDialog(props: Props) {
  /**
   * Custom Hooks
   */
  const { t } = useTranslation();
  const { isSmallScreen } = useScreenSize();

  /**
   * Callbacks
   */

  /**
   * Context
   */
  const { deleteCartItems } = useCartContext();
  const handleCancel = () => props.onClose(false);

  return (
    <Dialog open={props.open} maxWidth="sm" fullWidth onClose={props.onClose}>
      <DialogTitle>
        <Box color="primary.main">
          <Box pt={3}>
            {isSmallScreen ? (
              <Typography variant="h5" textAlign="center">
                {t('cart.removeAllFromCartDialog')}
              </Typography>
            ) : (
              <Typography variant="h5">
                {t('cart.removeAllFromCartDialog')}
              </Typography>
            )}
          </Box>
        </Box>
        <IconButton
          onClick={handleCancel}
          size="large"
          sx={(theme) => ({
            position: 'absolute',
            top: theme.spacing(3.5),
            right: theme.spacing(1)
          })}
        >
          <CloseIcon />
        </IconButton>
      </DialogTitle>
      <DialogContent>
        <Box py={2} color="primary.main">
          {isSmallScreen ? (
            <Typography component="h5" textAlign="center">
              {t('cart.removeAllFromCartDialogContent')}
            </Typography>
          ) : (
            <Typography component="h5">
              {t('cart.removeAllFromCartDialogContent')}
            </Typography>
          )}
        </Box>
        {isSmallScreen ? (
          <Box justifyContent="center">
            <Box justifyContent="center" display="flex">
              <Button
                sx={{ padding: '16px', mb: '8px' }}
                onClick={handleDeleteCart}
                data-testid="remove-all-items-dialog-confirm"
              >
                {t('cart.removeAllFromCartConfirm')}
              </Button>
            </Box>
            <Box justifyContent="center" display="flex">
              <Button
                onClick={handleCancel}
                variant="text"
                sx={{ padding: '16px' }}
              >
                {t('cart.removeAllFromCartCancel')}
              </Button>
            </Box>
          </Box>
        ) : (
          <Grid display="flex" justifyContent="right">
            <Grid paddingRight={3}>
              <Button
                onClick={handleCancel}
                variant="text"
                sx={{ padding: '16px', mr: 1 }}
                data-testid="remove-all-items-dialog-cancel"
              >
                {t('cart.removeAllFromCartCancel')}
              </Button>
            </Grid>
            <Grid>
              <Button
                onClick={handleDeleteCart}
                data-testid="remove-all-items-dialog-confirm"
                sx={{ padding: '16px', mr: 2 }}
              >
                {t('cart.removeAllFromCartConfirm')}
              </Button>
            </Grid>
          </Grid>
        )}
      </DialogContent>
    </Dialog>
  );

  function handleDeleteCart() {
    handleCancel();
    //Can't access button if cart is undefined
    deleteCartItems();
  }
}
export default RemoveCartItemsDialog;
