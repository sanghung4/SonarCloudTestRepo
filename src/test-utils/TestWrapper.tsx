import React, { JSXElementConstructor, ReactElement, ReactNode } from 'react';

import { ThemeProvider } from '@dialexa/reece-component-library';
import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { OktaAuth } from '@okta/okta-auth-js';
import { render as jestRender, RenderResult } from '@testing-library/react';
import { I18nextProvider } from 'react-i18next';
import { MemoryRouter } from 'react-router-dom';

import { UserContextType, AuthContext } from 'AuthProvider';
import { Permission } from 'common/PermissionRequired';
import i18n from 'test-utils/mockI18N';
import {
  SelectedAccountsContext,
  SelectedAccountsContextType
} from 'providers/SelectedAccountsProvider';
import { CartContext, CartContextType } from 'providers/CartProvider';
import {
  CheckoutContext,
  CheckoutContextType,
  initialCheckoutContextValues
} from 'Checkout/CheckoutProvider';

/**
 * NOTE:
 * To setup mocks follow the examples here:
 * https://www.apollographql.com/docs/react/development-testing/testing/
 *
 * To get a successful response in your test (not just the loading state)
 * you can do the following:
 *
 * it('Should render', async () => {
 *   const { container } = renderMockRouter(<MyComponent />, mocks);
 *
 *   await waitFor(() => new Promise((res) => setTimeout(res, 0)));
 *
 *   expect(container).toMatchSnapshot();
 * });
 *
 */

/******************************/
/* Types                      */
/******************************/
type Props = {
  children?: ReactNode;
  authConfig?: Partial<UserContextType>;
  selectedAccountsConfig?: Partial<SelectedAccountsContextType>;
  cartConfig?: Partial<CartContextType>;
  checkoutConfig?: Partial<CheckoutContextType>;
  mocks?: ReadonlyArray<MockedResponse>;
};

type WrapperProps<TKey extends keyof Props> = {
  value: Props[TKey];
  children: React.ReactNode;
};

/******************************/
/* Defaults                   */
/******************************/
const defaultAuth: UserContextType = {
  authState: {
    isAuthenticated: true
  },
  isLoggingOut: false,
  oktaAuth: new OktaAuth({ issuer: 'http://test.com' }),
  profile: {
    userId: 'testuser',
    permissions: Object.values(Permission),
    isEmployee: false,
    isVerified: false
  }
};

const defaultSelectedAccounts: SelectedAccountsContextType = {
  clearAccounts: () => {},
  updateAccounts: () => {},
  updateShippingBranchId: () => {},
  loading: false,
  selectedAccounts: {
    billTo: { erpAccountId: '' },
    shipTo: { erpAccountId: '' },
    shippingBranchId: ''
  },
  isMincron: false,
  isEclipse: false
};

const defaultCartContextValue: CartContextType = {
  addItemToCart: () => {},
  addItemsToCart: () => {},
  addAllListItemsToCart: () => {},
  cart: undefined,
  cartLoading: false,
  checkingOutWithQuote: false,
  contractBranch: undefined,
  clearContract: () => {},
  clearQuote: () => {},
  deleteCartItems: () => new Promise(() => {}),
  disableAddToCart: false,
  deleteItem: () => {},
  getUserCart: () => new Promise(() => {}),
  itemCount: 0,
  itemLoading: undefined,
  lineNotes: {},
  releaseContractToCart: () => {},
  setContract: () => {},
  setLineNotes: () => {},
  setQuoteId: () => {},
  setQuoteData: () => {},
  setQuoteShipToId: () => {},
  setPreviousCart: () => {},
  setSelectedBranch: () => {},
  updateCart: undefined,
  updateDelivery: undefined,
  updateItemQuantity: () => {},
  updateWillCall: undefined,
  updateWillCallBranch: undefined,
  refreshCart: undefined,
  isWillCall: false,
  deleteCreditCardFromCart: () => new Promise(() => {})
};

/******************************/
/* Wrappers                   */
/******************************/
function SelectedAccountsProviderWrapper({
  value,
  children
}: WrapperProps<'selectedAccountsConfig'>) {
  if (value) {
    return (
      <SelectedAccountsContext.Provider
        value={{ ...defaultSelectedAccounts, ...value }}
      >
        {children}
      </SelectedAccountsContext.Provider>
    );
  }

  return <>{children}</>;
}

function CartProviderWrapper({ value, children }: WrapperProps<'cartConfig'>) {
  if (value) {
    return (
      <CartContext.Provider value={{ ...defaultCartContextValue, ...value }}>
        {children}
      </CartContext.Provider>
    );
  }

  return <>{children}</>;
}

function CheckoutProviderWrapper({
  value,
  children
}: WrapperProps<'checkoutConfig'>) {
  if (value) {
    return (
      <CheckoutContext.Provider
        value={{ ...initialCheckoutContextValues, ...value }}
      >
        {children}
      </CheckoutContext.Provider>
    );
  }

  return <>{children}</>;
}

function TestWrapper(props: Props) {
  return (
    <ThemeProvider>
      <I18nextProvider i18n={i18n}>
        <AuthContext.Provider value={{ ...defaultAuth, ...props.authConfig }}>
          <MockedProvider mocks={props.mocks ?? []}>
            <SelectedAccountsProviderWrapper
              value={props.selectedAccountsConfig}
            >
              <CartProviderWrapper value={props.cartConfig}>
                <CheckoutProviderWrapper value={props.checkoutConfig}>
                  {props.children}
                </CheckoutProviderWrapper>
              </CartProviderWrapper>
            </SelectedAccountsProviderWrapper>
          </MockedProvider>
        </AuthContext.Provider>
      </I18nextProvider>
    </ThemeProvider>
  );
}

function jestRenderFn(wrapped: JSX.Element) {
  return jestRender(wrapped, {
    wrapper: MemoryRouter as
      | JSXElementConstructor<{
          children: ReactElement<any, string | JSXElementConstructor<any>>;
        }>
      | undefined
  });
}

export function render(component: ReactNode, config?: Props) {
  const wrapped = <TestWrapper {...config}>{component}</TestWrapper>;
  return jestRenderFn(wrapped);
}
export type TestRender = RenderResult<
  typeof import('@testing-library/dom/types/queries'),
  Element | DocumentFragment
>;

export function renderWithComponent(component: ReactNode, config?: Props) {
  const wrapped = <TestWrapper {...config}>{component}</TestWrapper>;
  return {
    render: jestRenderFn(wrapped),
    component: wrapped
  };
}

export default TestWrapper;
