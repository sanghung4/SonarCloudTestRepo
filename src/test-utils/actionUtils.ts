import { fireEvent, findByTestId, act } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

export const fillFormTextInput = async (testId: string, value: string) => {
  await act(async () => {
    const textInput = await findByTestId(document.body, `${testId}-input`);
    await fireEvent.change(textInput, { target: { value } });
  });
};

export const fillMaskedFormTextInput = async (
  testId: string,
  value: string
) => {
  await act(async () => {
    const textInput = await findByTestId(document.body, `${testId}-input`);
    await userEvent.type(textInput, value);
  });
};

export const clickButton = async (testId: string) => {
  await act(async () => {
    const button = await findByTestId(document.body, testId);
    await fireEvent.click(button);
  });
};

export const clickRadio = async (testId: string, optionIndex: number) => {
  await act(async () => {
    const radio = await findByTestId(
      document.body,
      `${testId}-radio-${optionIndex}`
    );
    await userEvent.click(radio);
  });
};

export const fillFormSelectInput = async (testId: string, value: string) => {
  await act(async () => {
    const selectInput = await findByTestId(document.body, `${testId}-input`);
    await fireEvent.change(selectInput, { target: { value } });
    await fireEvent.keyDown(window, { key: 'Enter' });
  });
};

export const focusInput = async (testId: string) => {
  await act(async () => {
    const input = await findByTestId(document.body, testId);
    await fireEvent.focus(input);
  });
};

export const changeInput = async (testId: string, value: string) => {
  await act(async () => {
    const textInput = await findByTestId(document.body, testId);
    await fireEvent.change(textInput, { target: { value } });
  });
};

export const blurInput = async (testId: string) => {
  await act(async () => {
    const input = await findByTestId(document.body, testId);
    await fireEvent.blur(input);
  });
};

export const keyboardDown = async (
  key: string,
  element?: Element | Node | Document | Window
) => {
  await act(async () => {
    await fireEvent.keyDown(element ?? window, { key });
  });
};
