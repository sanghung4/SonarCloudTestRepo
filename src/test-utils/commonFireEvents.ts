import { act, fireEvent } from '@testing-library/react';

export function actClickOnElement(element: HTMLElement) {
  act(() => {
    fireEvent.click(element);
  });
}
export function actUpdateInput(input: HTMLElement, value: string) {
  act(() => updateInput(input, value));
}
export function updateInput(input: HTMLElement, value: string) {
  fireEvent.focus(input);
  fireEvent.change(input, { target: { value } });
  fireEvent.blur(input);
}
