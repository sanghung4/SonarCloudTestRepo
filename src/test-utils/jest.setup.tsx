import React, { JSXElementConstructor, ReactElement, ReactNode } from "react";
import { MockedProvider, MockedResponse } from "@apollo/client/testing";
import { MemoryRouter } from "react-router-dom";
import { render as jestRender } from "@testing-library/react";

export const mockApollo = jest.mock(
  "@apollo/client/testing",
  () => MockedProvider
);

interface Props {
  children?: ReactNode;
  mocks?: ReadonlyArray<MockedResponse>;
}

function TestWrapper(props: Props) {
  return (
    <MockedProvider mocks={props.mocks ?? []}>{props.children}</MockedProvider>
  );
}

export function render(component: ReactNode, config?: Props) {
  const wrapped = <TestWrapper {...config}>{component}</TestWrapper>;
  return jestRender(wrapped, {
    wrapper: MemoryRouter as
      | JSXElementConstructor<{
          children: ReactElement<any, string | JSXElementConstructor<any>>;
        }>
      | undefined,
  });
}
