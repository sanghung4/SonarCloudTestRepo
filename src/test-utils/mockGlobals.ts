export function mockGoogle() {
  global.window.google = {
    maps: {
      // @ts-ignore
      places: {
        Autocomplete: jest.fn().mockImplementation(() => ({
          setFields: jest.fn(),
          setComponentRestrictions: jest.fn(),
          addListener: jest.fn()
        }))
      },
      Size: jest.fn(),
      Point: jest.fn()
    }
  };
}

export function mockGeolocation(lat: number, lng: number, err?: boolean) {
  // @ts-ignore
  global.navigator.geolocation = {
    getCurrentPosition: jest.fn().mockImplementation((onSuccess, onError) => {
      onSuccess({
        coords: {
          latitude: lat ?? 0,
          longitude: lng ?? 0
        }
      });
      if (err) {
        onError(new Error());
      }
    }),
    watchPosition: jest.fn().mockImplementation((onSuccess, onError) => {
      onSuccess({
        coords: {
          latitude: lat ?? 0,
          longitude: lng ?? 0
        }
      });
      if (err) {
        onError(new Error());
      }
    })
  };
}

export function mockLocalStorage() {
  // @ts-ignore
  global.window.localStorage = class mockLocalStorage {
    store: { [key: string]: string };
    constructor() {
      this.store = {};
    }
    clear() {
      this.store = {};
    }
    get length() {
      return Object.keys(this.store).length;
    }
    getItem(key: string) {
      return this.store[key] || null;
    }
    key(i: number) {
      return Object.keys(this.store)[i] || null;
    }
    removeItem(key: string) {
      delete this.store[key];
    }
    setItem(key: string, value: string) {
      this.store[key] = String(value);
    }
  };
}
