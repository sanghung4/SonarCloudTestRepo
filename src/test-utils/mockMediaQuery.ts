import mediaQuery from 'css-mediaquery';

function createMatchMedia(width: number) {
  return (query: string): MediaQueryList => ({
    matches: mediaQuery.match(query, { width }),
    media: query,
    onchange: null,
    addListener: () => jest.fn(),
    removeListener: () => jest.fn(),
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn()
  });
}

export const breakpoints = {
  mobile: 414,
  desktop: 1200
};

export const setBreakpoint = (breakpoint: 'mobile' | 'desktop' = 'desktop') => {
  Object.defineProperty(window, 'innerWidth', {
    writable: true,
    configurable: true,
    value: breakpoints[breakpoint]
  });
  window.dispatchEvent(new Event('resize'));

  window.matchMedia = createMatchMedia(window.innerWidth);
};
