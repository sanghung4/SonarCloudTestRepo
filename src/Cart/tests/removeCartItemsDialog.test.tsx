import { fireEvent } from '@testing-library/react';
import { noop } from 'lodash-es';

import RemoveCartItemsDialog from 'Cart/RemoveCartItemsDialog';
import { render } from 'test-utils/TestWrapper';

describe('Remove Cart Items Dialog', () => {
  it('Snapshot match with data', () => {
    const { baseElement } = render(
      <RemoveCartItemsDialog open={true} onClose={noop} />
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('Close Button click test', async () => {
    const testFunction = jest.fn();
    const { findByTestId } = render(
      <RemoveCartItemsDialog open={true} onClose={testFunction} />
    );
    const closeButtonTop = await findByTestId('remove-all-items-dialog-cancel');
    fireEvent.click(closeButtonTop);
    expect(testFunction).toBeCalledTimes(1);
  });
});
