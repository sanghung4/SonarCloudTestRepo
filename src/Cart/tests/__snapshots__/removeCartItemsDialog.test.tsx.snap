// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Remove Cart Items Dialog Snapshot match with data 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiModal-root MuiDialog-root css-zw3mfo-MuiModal-root-MuiDialog-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper css-15p4qre-MuiDialog-container"
      role="presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        aria-labelledby="mui-1"
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiDialog-paperFullWidth css-6gkcvy-MuiPaper-root-MuiDialog-paper"
        role="dialog"
      >
        <h2
          class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-tszvf-MuiTypography-root-MuiDialogTitle-root"
          id="mui-1"
        >
          <div
            class="MuiBox-root css-x84pqq"
          >
            <div
              class="MuiBox-root css-19n8dai"
            >
              <h5
                class="MuiTypography-root MuiTypography-h5 css-16xaac9-MuiTypography-root"
              >
                Are you sure you want to remove all items?
              </h5>
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeLarge css-dkqajj-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg>
              delete.svg
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </h2>
        <div
          class="MuiDialogContent-root css-ypiqx9-MuiDialogContent-root"
        >
          <div
            class="MuiBox-root css-11unscw"
          >
            <h5
              class="MuiTypography-root MuiTypography-body1 css-1k41qqb-MuiTypography-root"
            >
              Are you sure you want to remove all the items from the cart? This action can't be undone.
            </h5>
          </div>
          <div
            class="MuiGrid-root css-1ve4q92-MuiGrid-root"
          >
            <div
              class="MuiGrid-root css-4na9on-MuiGrid-root"
            >
              <button
                class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-disableElevation MuiButtonBase-root  css-yskaox-MuiButtonBase-root-MuiButton-root"
                data-testid="remove-all-items-dialog-cancel"
                tabindex="0"
                type="button"
              >
                No, cancel
              </button>
            </div>
            <div
              class="MuiGrid-root css-vj1n65-MuiGrid-root"
            >
              <button
                class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-disableElevation MuiButtonBase-root  css-1k977q5-MuiButtonBase-root-MuiButton-root"
                data-testid="remove-all-items-dialog-confirm"
                tabindex="0"
                type="button"
              >
                Yes, remove items
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
